### PR TITLE
[WIP] Python library to read GeoTrellis Avro records as Rasterio tile

### PR DESCRIPTION
**Pull request by shiraeeshi**
See https://github.com/geotrellis/geotrellis/issues/1403

### CODE REVIEW

**Comment by shiraeeshi** [comment]
Only ZSpatialKeyIndex is supported at this point. If the approach taken in the last commit is ok, then I'm going to work on all other KeyIndex types too. It's a lot of scala to python translation work, so I need some kind of confirmation.

**Comment by shiraeeshi** [comment]
I mean nobody is going to say: "You don't have to translate scala code to python, let's use Thrift (Py4J, JEP, Jython)".

**Comment by lossyrob** [comment]
My first comment is on project structure. Should we make this an actual python module, with a `setup.py`, that can be installed? Where should this be located? Perhaps a structure like having a `geotrellis-python` folder in the root of the repository, and then something like

```
.
├── CHANGES.txt
├── LICENSE.txt
├── MANIFEST.in
├── README.md
├── bin
│   └── geopy
├── geotrellis
│   ├── __init__.py
│   ├── io.py
└── setup.py
```

or a traditional python package structure underneath?

**Comment by lossyrob** [comment]
Otherwise the port portion makes me nervous, just as a matter of duplicate maintenance. That said, I don't see another way, and you're doing a good job with it. I would say a good project structure, and unit test coverage, would be ideal for this type of work.

**Comment by lossyrob** [comment]
@shiraeeshi really great stuff. Let us know how else we can support the work! Also you should pop into the Gitter channel and mention it, we could discuss there and get some more feedback from the members of the room.

File: geotrellis-python/geotrellis/python/util/doit.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from geotrellis.spark.io.file.FileValueReader import file_value_reader
+from geotrellis.spark.SpatialKey import SpatialKey
+from geotrellis.spark.LayerId import LayerId
+
+def doit(path_to_landsat, zoom = 0, col = 0, row = 0):
```
Comment (lossyrob) [inline geotrellis-python/geotrellis/python/util/doit.py]: What does this function do? Probably could use a more descriptive name :)

Comment (lossyrob) [inline geotrellis-python/geotrellis/python/util/doit.py]: Oh I see, this is the main entry point to test this.

**Comment by lossyrob** [comment]
Can we talk about the use case you're looking to support? Also wondering what the minimum viable support this should have in order to merge.

From what I can tell, you have ingested landsat images and you're looking to read out specific z/x/y tiles out of a file backend. Are you looking to expand support for other backends, e.g. Accumulo or S3?

Also, will this be expanded to read not just single tiles, but PySpark RDDs out?

If we can scope down these changes to a discrete purpose, perhaps with some unit tests and example of "this is how you would create a tile server using numpy with GeoTrellis ingested tiles" (for example), that would be a good goal to hit to get an initial commit of the python geotrellis code.

Really cool stuff!

**Comment by shiraeeshi** [comment]
I think that it's possible for this library to support all the backends GeoTrellis supports. The plan for this PR is to implement read/write operations using a file backend and PySpark RDDs.

I agree that there should be a way for developers to make sure that the code works by running tests and a sample server, I'll provide it in this PR.

**Comment by lossyrob** [comment]
@shiraeeshi wow, this PR is getting pretty big!

This is looking like a port of GeoTrellis to python. Is that the direction you are going? My concern is, of course, that maintaining a large project, and a port of that large project, is a lot of development burden.

It was my understanding at the outset of this work, that you'd be trying to read in rasterio (basically, numpy) rasters into PySpark RDDs, out of the Avro encodings that we had available. The porting over of the indexing and key bound logic made sense, because that is how you would come up with the query mechanism. I'm a bit surprised to see geotrellis-raster core functionality ported over.

To clarify, is the purpose of this work to be a python port of GeoTrellis as a whole?

**Comment by shiraeeshi** [comment]
It was hard for me decide what functionality to port and what not to, so I decided to create something that will work and will contain some extra code that will be later deleted. I started with io-related tests and tried to port all the code I could reach from it except third-party java code or geotrellis code tightly coupled with it.

Speaking about geotrellis-raster core functionality, I hesitated to code all the Tiles, but then decided to do so, here are some thoughts on this: a lot of avro schemas are named like `geotrellis.raster.SomethingArrayTile` so this library must have equivalents of all the Tile types. Tiles are tied with CellTypes, which are very important when deciding what value to use as nodata value. If I've created another python classes to represent all the Tiles with CellTypes, I mean created them from scratch without porting from scala, then they would become "black box" that nobody (except who created it) knows what's inside and nobody wants to deal with (and I would come up with something similar to original version anyway). But if if's just a port from scala, then everyone who wants to play with python version will know that the code is almost identical to scala code that he is familiar with already.

I think that all the methods in Tiles like `map`, `mapDouble`, `combine` was not necessary to port, because python version will work via rasterio, but it's hard to tell how rasterio will affect the code in tiles, so I ported that methods just in case.

**Comment by lossyrob** [comment]
Ok, thanks for the reasoning. The approach makes sense. I'm still nervous about having a lot of duplicated functionality in python of the Scala code...the worry be about the maintenance burden like I said. However this would be less of an issue if there is someone committed to maintaining it. Seems like you're pouring a lot of work into this, and I'm curious about what your future plans are for this? To be blunt, the "worst" case scenario is that we get a partial port of GeoTrellis to python, it goes unmaintained, people try to use it and it becomes a headache for people thinking the functionality would match the Scala functionality. Best case scenario is, together we figure out a good way to maintain it, port only what's necessary, and you take on the responsibility of porting fixes and updates to the python version (and plan to stick around for a bit ;)), and we get the luxury of using GeoTrellis in both Python and Scala (which would be awesome for sure). So, can you fill in some context around this work so I can get a better handle on where this is going?

Many thanks.

**Comment by shiraeeshi** [comment]
The job is almost done, I don't port anymore, 3 test specs are succeeding, now I'm working on making another 3 specs run. After that I'm planning to create a sample server.
I think that when all the tests run and sample server is ready will be a good time to think on how to make the library more compact and to create some boundaries on what to port to python, we will try to delete some code, tun tests, delete another portion, run the tests again, and so on.
I like the best case scenario. GeoTrellis is awesome project and I'd be happy and proud to be responsible for this python library.

**Comment by lossyrob** [comment]
> I like the best case scenario. GeoTrellis is awesome project and I'd be happy and proud to be responsible for this python library (having in mind that I don't own it and it's open source so everyone is free to work on it, I'm completely ok with that).

This makes me really happy!

I think that path forward sounds good. Would be great to hear when you get the other 3 specs running; either here or on the gitter channel. Feel free to discuss there to hash out any questions ping me and the other core devs for feedback.

**Comment by shiraeeshi** [comment]
OK, I'll report when this PR is ready, and I'll ask my questions here or in chat if I get stuck (I hope that I won't).

**Comment by shiraeeshi** [comment]
@lossyrob the tests are ready, there is also an instruction in readme on how to run them. Any feedback appreciated.

**Comment by shiraeeshi** [comment]
The sample server is ready. I believe this PR is ready to be merged.

**Comment by lossyrob** [comment]
I plan on reviewing this PR this week, stay tuned!

**Comment by lossyrob** [comment]
Sorry it's taking so long to get to this review; I've been pretty swamped. Just checking in to let you know it's still on the plate, hoping to get to it this week.

**Comment by shiraeeshi** [comment]
Can you make a fast review of one file at the moment?

**Comment by lossyrob** [comment]
@shiraeeshi sorry I didn't see your last comment until now! Which file did you want me to comment on specifically?

File: geotrellis-python/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+#### Dependencies
+
+- [Apache Avro](https://avro.apache.org/)
+- [Avro Json Serializer](https://github.com/linkedin/python-avro-json-serializer)
+- [NumPy](http://www.numpy.org/)
+- [rasterio](https://github.com/mapbox/rasterio)
+- [pytz](https://github.com/newvem/pytz)
+- [pypng](https://github.com/drj11/pypng)
+- [Spec](https://github.com/bitprophet/spec) for testing
```
Comment (lossyrob) [inline geotrellis-python/README.md]: When I do a `pip install spec` on my machine, I get `spec version 1.3.7`. When I run the `spec --with-inherited --with-tools-decorators-only` below, I get

```console
Usage: spec [options]

spec: error: no such option: --with-inherited
```

If I run spec without --with-inherited, I get 

```console
Usage: spec [options]

spec: error: no such option: --with-tools-decorators-only
```

if I run just `spec`, I get failures like
```console
TypeError: __init__() takes exactly 2 arguments (1 given)
```

Comment (lossyrob) [inline geotrellis-python/README.md]: Ah I see I didn't install the spec off your branch. Will try that now.

Comment (lossyrob) [inline geotrellis-python/README.md]: A bit of a pain to use a non-standard `spec` installation for this, and the change doesn't seem that big. Is it very necessary, or would we be able to use a standard install of spec by any chance?

File: geotrellis-python/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+#### Notes on compatibility
+
+The library will work with the following versions of dependencies (the command is `python setup.py install` under the main folder of the repo):
+- https://github.com/shiraeeshi/spec/tree/feature/options
+- https://github.com/shiraeeshi/python-avro-json-serializer/tree/feature/fullname
+- https://github.com/shiraeeshi/avro/tree/temp/with180version (The `setup.py` file resides in `lang/py` subdirectory)
```
Comment (lossyrob) [inline geotrellis-python/README.md]: Should this PR have a jira issue attached to it? Perhaps that will give it a better chance of being reviewed and merged https://issues.apache.org/jira/browse/AVRO/

File: geotrellis-python/README.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
+GeoTrellis uses [Apache Avro](https://avro.apache.org/) serialization when storing its tiled map layers. While GeoTrellis is a Scala project Avro provides libraries for multiple languages, including python. This project is to implement a python library that will read and write persisted GeoTrellis layers and present them as [NumPy](http://www.numpy.org/) datatypes that can be used by [rasterio](https://github.com/mapbox/rasterio). Once the basic translation is supported the next step would be to read GeoTrellis layer in a distributed Python Spark job and use rasterio to perform geospatial computations.
+
+Interoperability between GeoTrellis and rasterio will allow for complex, multi-stage geospatial workflows that do not suffer the usual performance penalties when working mixed Spark Scala/Python environment, expanding the utility of both libraries.
+
+#### Dependencies
```
Comment (lossyrob) [inline geotrellis-python/README.md]: Would be good to include a [requirements.txt](https://pip.readthedocs.io/en/1.1/requirements.html)

File: geotrellis-python/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+In order to run the tests, issue the following comand:
+
+`spec --with-inherited --with-tools-decorators-only` 
+
+from inside `geotrellis-python` directory.
+
+#### Sample server
+
+The landsat-django-server project (https://github.com/shiraeeshi/landsat-django-server) shows how to create a tile server with GeoTrellis ingested tiles.
```
Comment (lossyrob) [inline geotrellis-python/README.md]: Just tried to run this, got 

```console
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/rob/proj/gt/python/landsat-django-server/landsat/views.py", line 25, in zxy
    tile = reader.read(SpatialKey.fromTuple((x, y)))
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Geotrellis-0.1-py2.7.egg/geotrellis/spark/io/file/FileValueReader.py", line 49, in read
    return self.read_func(key)
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Geotrellis-0.1-py2.7.egg/geotrellis/spark/io/file/FileValueReader.py", line 35, in read_func
    recs = AvroEncoder.fromBinary(writer_schema, f.read(), codec = KeyValueRecordCodec(key_type, value_type))
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Geotrellis-0.1-py2.7.egg/geotrellis/spark/io/avro/AvroEncoder.py", line 54, in fromBinary
    result = codec.decode(dct)
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Geotrellis-0.1-py2.7.egg/geotrellis/spark/io/avro/codecs/KeyValueRecordCodec.py", line 26, in decode
    return map(lambda p: self.pairCodec.decode(p), dct["pairs"])
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Geotrellis-0.1-py2.7.egg/geotrellis/spark/io/avro/codecs/KeyValueRecordCodec.py", line 26, in <lambda>
    return map(lambda p: self.pairCodec.decode(p), dct["pairs"])
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Geotrellis-0.1-py2.7.egg/geotrellis/spark/io/avro/codecs/TupleCodec.py", line 38, in decode
    self.b.decode(dct["_2"]))
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Geotrellis-0.1-py2.7.egg/geotrellis/spark/io/avro/codecs/TileCodecs.py", line 416, in decode
    bands = map(self._tileUnionCodec.decode, dct["bands"])
  File "/usr/local/Cellar/python/2.7.9/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/Geotrellis-0.1-py2.7.egg/geotrellis/spark/io/avro/AvroUnionCodec.py", line 35, in decode
    schemaFullName = dct.schema.fullname
AttributeError: 'dict' object has no attribute 'schema'
```

Any ideas on what I'm doing wrong?

File: geotrellis-python/geotrellis/python/edited_avro/avro_builder.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
+from __future__ import absolute_import
```
Comment (lossyrob) [inline geotrellis-python/geotrellis/python/edited_avro/avro_builder.py]: What does `edited_avro` mean in this context?

Comment (lossyrob) [inline geotrellis-python/geotrellis/python/edited_avro/avro_builder.py]: Looks like it came from here, so I guess you mean an edited version of it https://github.com/Yelp/avro/blob/trunk/lang/py/src/avro/avro_builder.py

I see this file is not in trunk avro. Did they delete it in a later version? Is there another method of building up a schema that they recommend?

File: geotrellis-python/README.md (hunk 2 above)
Comment (shiraeeshi) [inline geotrellis-python/README.md]: Spec's standard behavior is pretty uncomfortable: it doesn't run inherited methods (which are widely used in geotrellis tests).

File: geotrellis-python/README.md (hunk 3 above)
Comment (shiraeeshi) [inline geotrellis-python/README.md]: Good point, there is an issue in jira, which seems relevant (https://issues.apache.org/jira/browse/AVRO-1777), I've commented there.

File: geotrellis-python/geotrellis/python/edited_avro/avro_builder.py (hunk 6 above)
Comment (shiraeeshi) [inline geotrellis-python/geotrellis/python/edited_avro/avro_builder.py]: There is an issue in jira (https://issues.apache.org/jira/browse/AVRO-1690), the avro team didn't accept the PR and didn't comment.

File: geotrellis-python/README.md (hunk 5 above)
Comment (shiraeeshi) [inline geotrellis-python/README.md]: Looks like you're using the standard version avro instead of my branch.

**Comment by shiraeeshi** [comment]
> @shiraeeshi sorry I didn't see your last comment until now! Which file did you want me to comment on specifically?

I thought about AvroEncoder, but don't feel obliged to review that particular file, I just figured it's good place to start from.

**Comment by lossyrob** [comment]
Hey @shiraeeshi, I pinged you on Gitter; leaving a message here in case you didn't see that.

**Comment by lossyrob** [comment]
I'm closing this PR as it seems to have gone a bit stale. This will be used to seed some Python bindings work we have coming up. We're looking to get python integration done by Spring 2017.

**Pull request closed without merging**